# Heatmaps
Are really cool!

###### 

#### They actully show 3 features together.

- **X axis** → Continuous / Categorical Feature
- **Y axis** → Continuous / Categorical Feature
- **Map - Grid** → Continuours Feature

In [1]:
# Back to school - heatmaps are NOT complex. So they are available in the go.
import pandas as pd
import plotly.offline as pyo
import plotly.graph_objs as go

In [22]:
df = pd.read_csv("../../DATA/Data/2010SantaBarbaraCA.csv")
df

LST_DATE      DAY LST_TIME  T_HR_AVG
0    20100601  TUESDAY     0:00      12.7
1    20100601  TUESDAY     1:00      12.7
2    20100601  TUESDAY     2:00      12.3
3    20100601  TUESDAY     3:00      12.5
4    20100601  TUESDAY     4:00      12.7
..        ...      ...      ...       ...
163  20100607   MONDAY    19:00      15.6
164  20100607   MONDAY    20:00      14.8
165  20100607   MONDAY    21:00      14.3
166  20100607   MONDAY    22:00      14.4
167  20100607   MONDAY    23:00      14.6

[168 rows x 4 columns]

## Something is... a bit weird here. In the syntax. We'll see

###### 

# First — Simple plot (sorry Map)

In [4]:
plot = [go.Heatmap(x=df["DAY"],
                   y=df["LST_TIME"],
                   z=df["T_HR_AVG"])]
pyo.plot(plot)

'file://C:\\Users\\GIGABYTE\\Desktop\\Plot.ly\\1. Plotly — Basics\\7. Heatmaps\\temp-plot.html'

# Second — With some color...

In [6]:
plot = [go.Heatmap(x=df["DAY"],
                   y=df["LST_TIME"],
                   z=df["T_HR_AVG"], 
                  colorscale="Blues")]
            ### POI ↑ POI ###
pyo.plot(plot)

'file://C:\\Users\\GIGABYTE\\Desktop\\Plot.ly\\1. Plotly — Basics\\7. Heatmaps\\temp-plot.html'

Now, the **PROBLEM**, <br>
As you can see from the syntax, it lets use to pass `x, y, z` together. So that means we don't have to make a table like structure or in other words, we don't have to make the `pivot table` with hands and then do the heatmap call like we are used to.

But what if, we have same rows repeating? Will it take mean, or median? or what for the Z?

In [34]:
df_redundant = df.copy()

In [36]:
df_redundant.head(3)

LST_DATE      DAY LST_TIME  T_HR_AVG
0  20100601  TUESDAY     0:00      12.7
1  20100601  TUESDAY     1:00      12.7
2  20100601  TUESDAY     2:00      12.3

In [38]:
df_redundant.iloc[1, 2] = '0:00'
df_redundant.iloc[1, 3] = '20'

In [39]:
df_redundant.head(3)

LST_DATE      DAY LST_TIME T_HR_AVG
0  20100601  TUESDAY     0:00     12.7
1  20100601  TUESDAY     0:00       20
2  20100601  TUESDAY     2:00     12.3

In [40]:
# Checking that out.
plot = [go.Heatmap(x=df_redundant["DAY"],
                   y=df_redundant["LST_TIME"],
                   z=df_redundant["T_HR_AVG"])]
pyo.plot(plot)

'file://C:\\Users\\GIGABYTE\\Desktop\\Plot.ly\\1. Plotly — Basics\\7. Heatmaps\\temp-plot.html'

## WTFun! It took the second one.
Means, as we have made the second row same as the first, but with the change at 20. It took 20 instead of 12.7 that means **no aggregation is performed**.

And in addition, as we did that, we removed TUESDAY - 1:00 from the data so it took NaN there.

# So tha question
> What if I have to perform some aggfunc with pivot table and then?

**Answer:** We simply have to make it convert to the X, Y, Z format. <br>
Shown in ↓

In [69]:
step_1 = df.pivot_table(index="DAY", columns="LST_TIME", values="T_HR_AVG").T
step_1

DAY       FRIDAY  MONDAY  SATURDAY  SUNDAY  THURSDAY  TUESDAY  WEDNESDAY
LST_TIME                                                                
0:00        13.8    15.1      14.2    15.0      12.1     12.7       11.3
10:00       15.9    15.6      16.2    15.8      14.5     13.2       14.9
11:00       16.3    15.8      17.2    16.3      14.9     14.1       16.0
12:00       16.5    16.5      17.3    17.4      15.7     14.8       16.5
13:00       16.8    16.5      17.2    17.6      15.6     15.3       17.0
14:00       17.3    16.8      17.0    17.7      15.5     15.6       17.1
15:00       17.0    17.5      16.9    17.4      15.8     15.8       16.9
16:00       16.3    17.7      16.2    16.4      15.6     16.0       16.4
17:00       15.6    17.6      15.5    16.2      14.9     15.5       15.9
18:00       15.3    16.5      15.3    16.0      14.9     14.9       15.4
19:00       15.0    15.6      14.6    15.0      14.5     14.4       14.4
1:00        13.0    14.9      14.0    15.0      11.9     12.7       11.1
20:00       14.4    14.8      15.0    13.9      13.8     13.6       13.6
21:00       14.3    14.3      15.2    13.7      13.7     13.1       13.7
22:00       14.2    14.4      15.0    14.6      13.6     12.3       12.8
23:00       14.1    14.6      15.1    15.0      13.7     11.9       12.8
2:00        12.7    14.8      13.7    15.0      11.6     12.3       10.8
3:00        13.3    14.8      13.7    15.1      11.5     12.5       10.5
4:00        13.6    14.9      13.7    15.2      11.2     12.7       10.4
5:00        13.6    14.9      13.7    15.2      12.4     12.7       10.3
6:00        13.7    14.8      13.8    15.2      12.9     13.0       11.8
7:00        14.1    14.9      14.2    15.3      13.7     12.7       13.7
8:00        14.7    15.1      14.6    15.3      14.0     12.8       14.5
9:00        14.8    15.3      15.1    15.3      14.1     12.7       14.3

In [70]:
step_2 = step_1.stack()
step_2

LST_TIME  DAY      
0:00      FRIDAY       13.8
          MONDAY       15.1
          SATURDAY     14.2
          SUNDAY       15.0
          THURSDAY     12.1
                       ... 
9:00      SATURDAY     15.1
          SUNDAY       15.3
          THURSDAY     14.1
          TUESDAY      12.7
          WEDNESDAY    14.3
Length: 168, dtype: float64

In [71]:
step_3 = step_2.reset_index()
step_3

LST_TIME        DAY     0
0       0:00     FRIDAY  13.8
1       0:00     MONDAY  15.1
2       0:00   SATURDAY  14.2
3       0:00     SUNDAY  15.0
4       0:00   THURSDAY  12.1
..       ...        ...   ...
163     9:00   SATURDAY  15.1
164     9:00     SUNDAY  15.3
165     9:00   THURSDAY  14.1
166     9:00    TUESDAY  12.7
167     9:00  WEDNESDAY  14.3

[168 rows x 3 columns]

In [72]:
plot = [go.Heatmap(x=step_3["DAY"],
                   y=step_3["LST_TIME"],
                   z=step_3[0])]
pyo.plot(plot)

'file://C:\\Users\\GIGABYTE\\Desktop\\Plot.ly\\1. Plotly — Basics\\7. Heatmaps\\temp-plot.html'

##### 

### So... let's explore more.

###### 

# SUBPLOTS in plotly

In [73]:
df1 = pd.read_csv("../../DATA/Data/2010SantaBarbaraCA.csv")
df2 = pd.read_csv("../../DATA/Data/2010SitKaak.csv")
df3 = pd.read_csv("../../DATA/Data/2010YumaAz.csv")

In [75]:
trace_1 = go.Heatmap(x=df1["DAY"],
                       y=df1["LST_TIME"],
                       z=df1["T_HR_AVG"])

trace_2 = go.Heatmap(x=df2["DAY"],
                       y=df2["LST_TIME"],
                       z=df2["T_HR_AVG"])

trace_3 = go.Heatmap(x=df3["DAY"],
                       y=df3["LST_TIME"],
                       z=df3["T_HR_AVG"])

In [76]:
# Wrong Way (but just tried)

traces = [trace_1, trace_2, trace_3]
pyo.plot(traces)

'file://C:\\Users\\GIGABYTE\\Desktop\\Plot.ly\\1. Plotly — Basics\\7. Heatmaps\\temp-plot.html'

#### Oops! Things are overlapping

Because here we **NEED TO CREATE SUBPLOTS**.

In [77]:
# Right Way
from plotly import tools

In [87]:
                    ## POI ↓ POI ###
    
fig = tools.make_subplots(rows=1, cols=3,
                          subplot_titles=["Sitka", "AK", "AZ"], # ← Also note this
                          shared_yaxes=True) # ← Also note this

This is the format of your plot grid:
[ (1,1) x1,y1 ]  [ (1,2) x2,y1 ]  [ (1,3) x3,y1 ]



Now, our subplot is ready. Now we need to point plotly that WHERE EACH OF THE PLOT SHOULD GO.

In [89]:
fig.append_trace(trace_1, 1, 1)
fig.append_trace(trace_2, 1, 2)
fig.append_trace(trace_3, 1, 3)

In [91]:
pyo.plot(fig)

'file://C:\\Users\\GIGABYTE\\Desktop\\Plot.ly\\1. Plotly — Basics\\7. Heatmaps\\temp-plot.html'

**NOTE:** If you would have passed FIG without appening any single trace, then the error will be there.

# COOL! But still...
The colorbar is `overlapping`.  To fix that, we have to give the same color range to all plots.

```python
trace_1 = go.Heatmap(x=..., y=..., z=...,
                    zmin=0, zmax=20)
```
So that will make all colorbar to the same range—rather making seperate one.

In [92]:
zminmax = dict(zmin=0, zmax=20)

trace_1 = go.Heatmap(x=df1["DAY"],
                       y=df1["LST_TIME"],
                       z=df1["T_HR_AVG"], **zminmax)

trace_2 = go.Heatmap(x=df2["DAY"],
                       y=df2["LST_TIME"],
                       z=df2["T_HR_AVG"], **zminmax)

trace_3 = go.Heatmap(x=df3["DAY"],
                       y=df3["LST_TIME"],
                       z=df3["T_HR_AVG"], **zminmax)

fig = tools.make_subplots(1, 3, shared_yaxes=True, subplot_titles=["Kaka1", "Kaka2", "Kaka3"])
fig.append_trace(trace_1, 1, 1)
fig.append_trace(trace_2, 1, 2)
fig.append_trace(trace_3, 1, 3)

pyo.plot(fig)

This is the format of your plot grid:
[ (1,1) x1,y1 ]  [ (1,2) x2,y1 ]  [ (1,3) x3,y1 ]



'file://C:\\Users\\GIGABYTE\\Desktop\\Plot.ly\\1. Plotly — Basics\\7. Heatmaps\\temp-plot.html'

# COOLER!

# 

# More to Note... 
See, there `fig`. We have been used to define the `fig` only with:
```python
fig = go.Figure(data=data, layout=layout)
```
Right? But here, we don't have to do that.
```python
fig = tools.make_subplots()
# This ↑ fig
```
**IS THE SAME** fig as we would have made. So, as that fig involves / binds `data=data` and `layout=layout` togehter, we can access the layout from there itself.

#### Like if we want to set the MAJOR TITLE

In [96]:
fig["layout"]["title"] = "Major Title Good One"
pyo.plot(fig)

##### Got the title! Whoo!
In other words, when we create subplotsm **We can't make the `layout` variable** ourselvs. Instead we have to access it from the fig. 

That's not bad.

# That's it!
Nextly another practice exercise!